cleanrl/ppo_dna_atari_envpool.py

```diff
@@ -0,0 +1,468 @@
+        return (
+            observations,
+            rewards,
+            dones,
+            infos,
+        )
+
+
+def compute_advantages(rewards, dones, values, next_done, next_value, gamma, gae_lambda):
+    total_steps = len(rewards)
+    if gae_lambda > 0:
+        advantages = torch.zeros_like(rewards)
+        lastgaelam = 0
+        for t in reversed(range(total_steps)):
+            if t == total_steps - 1:
+                nextnonterminal = 1.0 - next_done
+                nextvalues = next_value
+            else:
+                nextnonterminal = 1.0 - dones[t + 1]
+                nextvalues = values[t + 1]
+            delta = rewards[t] + gamma * nextvalues * nextnonterminal - values[t]
+            advantages[t] = lastgaelam = delta + gamma * gae_lambda * nextnonterminal * lastgaelam
+        returns = advantages + values
+    else:
+        returns = torch.zeros_like(rewards)
+        for t in reversed(range(total_steps)):
+            if t == total_steps - 1:
+                nextnonterminal = 1.0 - next_done
+                next_return = next_value
+            else:
+                nextnonterminal = 1.0 - dones[t + 1]
+                next_return = returns[t + 1]
+            returns[t] = rewards[t] + gamma * nextnonterminal * next_return
+        advantages = returns - values
+    return advantages, returns
+
+
+def layer_init(layer, std=np.sqrt(2), bias_const=0.0):
+    torch.nn.init.orthogonal_(layer.weight, std)
+    torch.nn.init.constant_(layer.bias, bias_const)
+    return layer
+
+
+class Agent(nn.Module):
+    def __init__(self, envs):
+        super().__init__()
+        self.network = nn.Sequential(
+            layer_init(nn.Conv2d(4, 32, 8, stride=4)),
+            nn.ReLU(),
+            layer_init(nn.Conv2d(32, 64, 4, stride=2)),
+            nn.ReLU(),
+            layer_init(nn.Conv2d(64, 64, 3, stride=1)),
+            nn.ReLU(),
+            nn.Flatten(),
+            layer_init(nn.Linear(64 * 7 * 7, 512)),
+            nn.ReLU(),
+        )
+        self.actor = layer_init(nn.Linear(512, envs.single_action_space.n), std=0.01)
+        self.critic = layer_init(nn.Linear(512, 1), std=1)
+
+    def forward(self, x):
+        """Run policy and value functions."""
+        hidden = self.network(x)
+        logits = self.actor(hidden)
+        value = self.critic(hidden).squeeze(-1)
+        return logits, value
+
+    @staticmethod
+    def sample_action(logits):
+        probs = Categorical(logits=logits)
+        action = probs.sample()
+        return action, probs.log_prob(action)
+
+    @staticmethod
+    def action_log_prob(logits, action):
+        probs = Categorical(logits=logits)
+        return probs.log_prob(action), probs.entropy()
+
+    @staticmethod
+    def action_kl_divergence(old_logits, new_logits):
+        old_probs = Categorical(logits=old_logits)
+        new_probs = Categorical(logits=new_logits)
+        return kl_divergence(old_probs, new_probs).mean()
+
+
+def main():
```

Comment:
Out of curiosity, what is the purpose of using the `main()` function? We loss a bit of flexibility in debug when we do something like 

```
python ppo_dna_atari_envpool.py

# run it for 10 seconds then ctrl + c to print variables
```

Reply:
yeah, it's because I extracted `compute_advantages` as a separate function, but then noticed that it was using values of global variables defined later in the module, which is not what I intended. It's essentially a safeguard to remove all global variables with generic names like `reward` or `value` etc.

Reply:
Of course, then I fixed the compute_advantages arguments list to include all used variables, so it's no longer using global variables (I think), but in general, I feel it would be nicer that the program fails with an undefined variable instead of silently using a global variable defined somewhere else

However, I realize that this is not a problem if everything is in global scope and there are no functions (which pretty much is the cleanRL style), but I think that the compute_advantages function is necessary in this case, and that's why I opted for increased safety against (future) bugs, at the cost of decreased interactive convenience

Personally, I edit the file locally to replace `def main():` with `if 1:` for a smooth interactive experience, if needed. But I'm happy to revert this change as well if you think so; lmk!

Reply:
If we'd like to have all the variables globally, its possible to add two extra lines in the original GAE calculation code to have the two separate advantage calculation steps. I think almost everything else remains same.

```
delta_policy = rewards[t] + gamma * nextvalues * nextnonterminal - values[t]
advantages_policy[t] = lastgaelam = delta_policy + gamma * policy_lambda * nextnonterminal * lastgaelam
delta_value = rewards[t] + gamma * nextvalues * nextnonterminal - values[t]
advantages_value[t] = lastgaelam = delta_value + gamma * value_lambda * nextnonterminal * lastgaelam
``` 

Generally I prefer functions more but cleanrl avoids them so this can be an alternative.

Reply:
Thanks @Dipamc77 for the suggestion. In this case though, using `compute_advantages` seems pretty clean to me, so I am ok with it :)

Reply:
>  It's essentially a safeguard to remove all global variables with generic names like reward or value etc.

This is a good point. I was thinking of using a pre-commit hook like pylint to detect this kind of situations. For now, let's revert back to not using the `main()` function to be consistent w/ other scripts

Reply:
done